snuba_sdk/visitors.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+from datetime import date, datetime
+from typing import Generic, TypeVar
+
+from snuba_sdk.expressions import (
+    Column,
+    Expression,
+    Function,
+    Granularity,
+    InvalidExpression,
+    Limit,
+    Offset,
+    Scalar,
+    ScalarType,
+)
+from snuba_sdk.entity import Entity
+from snuba_sdk.conditions import Condition
+
+
+# validation regexes
+unescaped_quotes = re.compile(r"(?<!\\)'")
+unescaped_newline = re.compile(r"(?<!\\)\n")
+
+
+def _stringify_scalar(value: ScalarType) -> str:
+    if value is None:
+        return "NULL"
+    elif isinstance(value, bool):
+        return "TRUE" if value else "FALSE"
+    if isinstance(value, (str, bytes)):
+        if isinstance(value, bytes):
+            decoded = value.decode()
+        else:
+            decoded = value
+
+        decoded = unescaped_quotes.sub("\\'", decoded)
+        decoded = unescaped_newline.sub("\\\\n", decoded)
+        return f"'{decoded}'"
+    elif isinstance(value, (int, float)):
+        return f"{value}"
+    elif isinstance(value, datetime):
+        # Snuba expects naive UTC datetimes, so convert to that
+        if value.tzinfo is not None:
+            delta = value.utcoffset()
+            assert delta is not None
+            value = value - delta
+            value = value.replace(tzinfo=None)
+        return f"toDateTime('{value.isoformat()}')"
+    elif isinstance(value, date):
+        return f"toDateTime('{value.isoformat()}')"
+
+    raise InvalidExpression(f"'{value}' is not a valid scalar")
+
+
+TVisited = TypeVar("TVisited")
+
+
+class Visitor(ABC, Generic[TVisited]):
+    def visit(self, node: Expression) -> TVisited:
+        if isinstance(node, Column):
+            return self._visit_column(node)
+        elif isinstance(node, Function):
+            return self._visit_function(node)
+        elif isinstance(node, Entity):
+            return self._visit_entity(node)
+        elif isinstance(node, Condition):
+            return self._visit_condition(node)
+        elif isinstance(node, Limit):
+            return self._visit_int_literal(node.limit)
+        elif isinstance(node, Offset):
+            return self._visit_int_literal(node.offset)
+        elif isinstance(node, Granularity):
+            return self._visit_int_literal(node.granularity)
+
+        raise NotImplementedError
```

Comment:
This explains all the details about the exhaustive checks https://hakibenita.com/python-mypy-exhaustive-checking#type-narrowing-in-mypy

Reply:
Oh whoops yes I read that article and changed this but forgot to push the change. I'll update.

Reply:
I wasn't able to use the `assert_never` function because this function takes an Expression, and mypy isn't smart enough to know that I have checked for all subclasses of Expression, which is itself an ABC. I don't really want to have a giant `Union[Column, Function...]` in the argument types either.

Reply:
ok, I give up, this is python telling us: "I don't have pattern matching. Stop trying to use it."